src/libcollections/str.rs

```diff
@@ -273,7 +273,7 @@ impl<'a> Iterator for Recompositions<'a> {
         loop {
             match self.state {
                 Composing => {
-                    for ch in self.iter {
+                    while let Some(ch) = self.iter.next() {
```

Comment:
Could some loops like this be altered to:

```rust
for ch in self.iter.by_ref() { /* ... */ }
```

That may avoid consuming the entire iterator.